File: 2015/day11.js
const hasStraight = arr => (
  arr.some((n, i, chars) => (
    i > 1 && n - 1 === chars[i - 1] && n - 2 === chars[i - 2]
  ))
)

const illegalCodes = ['i', 'o', 'l'].map(letter => letter.charCodeAt(0))
const noIllegals = arr => (
  illegalCodes.every(letter => !arr.includes(letter))
)

const twoPair = arr => (
  arr.some((n, i, chars) => (
    i > 0 && n === chars[i - 1] && arr.some((m, j) => (
      j > 2 && m !== n && m === chars[j - 1]
    ))
  ))
)

const toCharCode = str => (
  str.split('').map(char => char.charCodeAt(0)).filter(char => char > 96)
)
const fromCharCode = arr => (
  arr.map(String.fromCharCode).join('')
)

const firstTest = toCharCode('hijklmmn')
console.log(hasStraight(firstTest))
console.log(!noIllegals(firstTest))

const secondTest = toCharCode('abbceffg')
console.log(twoPair(secondTest))
console.log(!hasStraight(secondTest))

const legal = codes => noIllegals(codes) && hasStraight(codes) && twoPair(codes)

const increment = charCodeArr => {
  const lastPosition = charCodeArr.length - 1
  const allButLast = charCodeArr.slice(0, lastPosition)
  const last = charCodeArr[lastPosition]
  if (last === 122) { //'z'.charCodeAt(0)
    return increment(allButLast).concat(97) //'a'.charCodeAt(0)
  } else {
    return allButLast.concat(last + 1)
  }
}

//const pwd = 'abcdefgh'
const pwd = 'hepxxyzz'
const nextPassword = password => {
  let code = toCharCode(password)
  while (!legal(code)) {
    code = increment(code)
  }
  return fromCharCode(code)
}
const nextPwd = nextPassword(pwd)
console.log(nextPwd)
console.log(nextPassword(fromCharCode(increment(toCharCode(nextPassword(pwd))))))

// for some reason, I couldn't automate the incrementing between passwords
// without filtering out spurious char codes - how did those get there?
// 

File: 2015/day11.rb
input = 'hepxcrrq'

class Password
  def initialize(str)
    @pwd = str.split('')
  end

  def has_straight?
    @pwd.each_with_index.any?{|char, i| @pwd[i + 1] == char.succ && @pwd[i + 2] == char.succ.succ }
  end

  def not_confusing?
    @pwd.none?{|char| %w(i o l).include? char }
  end

  def two_pair?
    ('a'..'z').any? do |char| 
      pwd.include?(char * 2) && ('a'..'z').reject{|c| c == char }.any?{|c| pwd.include?(c * 2)}
    end
  end

  def pwd
    @pwd.join
  end

  def valid?
    has_straight? && not_confusing? && two_pair?
  end

  def next
    loop do
      @pwd = pwd.succ.split('')
      break if valid?
    end
  end
end

pwd = Password.new(input)
pwd.next
puts pwd.pwd
pwd.next
puts pwd.pwd
